File: 0046-permutations/0046-permutations.cpp
class Solution {

public:
    void auxPermute(int ind, int n, vector<int>& nums, vector<vector<int>>& ans,
                    map<int, int>& mp, vector<int>& ds) {

        if (ds.size() == n) {
            ans.push_back(ds);
            return;
        }

        for (int i = ind; i < n; i++) {

            if (!mp[i]) {

                mp[i] = 1;
                ds.push_back(nums[i]);
                auxPermute(ind, n, nums, ans, mp, ds);
                ds.pop_back();
                mp[i] = 0;
            }
        }
    }

public:
    vector<vector<int>> permute(vector<int>& nums) {

        vector<vector<int>> ans;
        vector<int> ds;
        map<int, int> mp;

        auxPermute(0, nums.size(), nums, ans, mp, ds);
        return ans;
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        auxPermute(0, nums.length, nums, res);
        return res;
    }

    public static void auxPermute(int ind, int n, int ds[], List<List<Integer>> res) {
        if (ind == n) {
            List<Integer> temp = new ArrayList<>();
            for (int num : ds)
                temp.add(num);
            res.add(temp);
        }

        for (int i = ind; i < n; i++) {
            swap(ds, i, ind);
            auxPermute(ind + 1, n, ds, res);
            swap(ds, ind, i);

        }
    }

    public static void swap(int[] ds, int i, int j) {
        int temp = ds[i];
        ds[i] = ds[j];
        ds[j] = temp;
    }

}